### ISSUE TITLE
Feature for consideration: implement 'best image' detection

### ISSUE THREAD

**Issue by jayaddison** [hhursev/recipe-scrapers#883]
This feature could be up for discussion - there are a few factors to consider.

Currently, when asking `recipe-scrapers` for the `image` field for a scraped recipe, and when a recipe page provides a list of images, [we'll return the first one](https://github.com/hhursev/recipe-scrapers/blob/bbf28484eeeb918e97d6815860f74d22a172b310/recipe_scrapers/_schemaorg.py#L147-L149).  There's no guarantee that'll be a good image to display; it's subjective and also it depends on the ordering of the images that the recipe page provides.

The `mealie` project have improved on this by adding a feature that implements best-image-detection by -- roughly speaking -- downloading the images and considering the largest one to be the best one to display.

If we implemented a similar algorithm within the library here, `mealie` could use the standard `image` field access method instead of custom logic, and we'd likely improve the recipe data experience for our other users too.

Note: we should be careful to comply with software licensing here; `mealie` is AGPLv3 licensed, so if we borrow that code, we'll want to provide fixes and improvements to it back to `mealie`, and should let them know about those.  In practice it seems like `mealie` _might_ be able to completely drop the code and re-use it from `recipe-scrapers` if it became part of our codebase - but I'm not a lawyer and so I think we'd want to make sure everyone's informed and comfortable with the approach for that.  Another option would be for someone who hasn't read the `mealie` codebase to implement a fresh, similar algorithm for us in `recipe-scrapers` without inspecting the `mealie` code, in which case licensing isn't so much of a concern.  (although that makes me wonder: how could our reviewers know that the implementor _hasn't_ read that code?)

And declaring a conflict-of-interests of sorts: I won't want to use this feature in my projects, because I generally want to minimize the amount of network bandwidth between client and recipe website.  Even so, I think it's a good feature and probably valuable for most users, who are likely only to request a few recipes, and not generate or require much bandwidth for that.  For these reasons, I'd like if this could be a configuration setting or optional feature -- but I'm OK with it being the default, personally.

cc @hay-kot @cadamswaite

(maybe I'm being pedantic about the software licensing concerns; my internal mental model is that it's worth being careful about those)

**Comment by hay-kot** [comment]
> The mealie project have improved on this by adding a feature that implements best-image-detection by -- roughly speaking -- downloading the images and considering the largest one to be the best one to display.

To clarify, and provide some additional context - we don't download the images, we issues a HEAD request and then try to read the Content-Length header and uses that to base the choice off of (you probably already knew this, but I wanted to clarify in case anyone else stumbles upon this issue). 

HOWEVER, as always in this space there are exceptions where the sites don't set a Content-Length header and I've had to deal with headaches trying to figure out _why_ it was crashing. I ended up having to implement a strategy that would first try the HEAD, then it would issue a GET request, and then fall back to just using the first URL provided we didn't get a 4xx from the implementation. This code isn't public yet and it's written in Go, so I don't know how helpful it would be for this implementation.

I'd also _highly_ suggest implementing this in an async manner, it's significantly faster than running requests synchronously. In my testing, running these synchronously could add 2-3 seconds or more in some cases with many images.

> Note: we should be careful to comply with software licensing here; mealie is AGPLv3 licensed, so if we borrow that code, we'll want to provide fixes and improvements to it back to mealie, 

ultimately, Mealie does a lot of "cleaning" of data that I think can be upstreamed here. As the primary maintainer, I'm okay with that work getting upstreamed and re-licensed to MIT. There is a slight problem here in that we heavily use pattern matching, a Python 3.10 feature, so depending on the projects willingness to support 3.9, that code would have to be reworked, and it then becomes a very ugly mess :(

**Comment by jayaddison** [comment]
> > Note: we should be careful to comply with software licensing here; mealie is AGPLv3 licensed, so if we borrow that code, we'll want to provide fixes and improvements to it back to mealie,
> 
> ultimately, Mealie does a lot of "cleaning" of data that I think can be upstreamed here. As the primary maintainer, I'm okay with that work getting upstreamed and re-licensed to MIT.

To be strict I think we'd also need the approval of the original author(s) for each part of the code migrated across (in this case, that's @cadamswaite) - thank you though, glad for the support.

This isn't a situation I've personally experienced before - migrating code from one place to another across a software license boundary - as far as I can remember, so I'll try to read a bit more about it.

> There is a slight problem here in that we heavily use pattern matching, a Python 3.10 feature, so depending on the projects willingness to support 3.9, that code would have to be reworked, and it then becomes a very ugly mess :(

Ok; no hurry I suppose.  We did only drop py37 support a few weeks ago, though :grin: 

If there are any particular cleanup tasks that are in `mealie` that you think would be particularly good to migrate here, perhaps we could open issue threads for those?

**Comment by cadamswaite** [comment]
I waive all rights to the code I've contributed to mealie.
Feel free to take as much/little as you want without attribution

**Comment by jayaddison** [comment]
I'm about to read some of this document that looks like a good resource: https://softwarefreedom.org/resources/2008/foss-primer.html

Odd though it might seem, I think we might be on stronger grounds if we do attribute the source of the code.  And if I remember correctly, some jurisdictions don't allow waiving rights to creative works.  So, I guess I'm being nitpicky, but deciding whether you're OK with your code being relicensed to MIT for the purpose of inclusion in `recipe-scrapers` could be the question.

**Comment by cadamswaite** [comment]
Feel free to use it under any licence.

As far as I'm concerned, the code I contributed to mealie is public domain. If it needs a licence I choose [WTFPL](https://choosealicense.com/licenses/wtfpl/)

**Comment by jayaddison** [comment]
From my understanding of licensing and contributions, and trying to imagine the point of view of a software licensing attorney, it seems fairly clear (unless you have a prior / private agreement with the mealie project, or had published the code elsewhere under a public domain license) to me that the best-image-detection code is licensed under AGPLv3 as part of that project.

That said, I don't think there'd be anything preventing someone who authored some code in an individual capacity from making an equivalent contribution of similar/equivalent code to another project that uses different licensing terms.

I have to rule myself out from implementing this I think, because I did read the `mealie` code relating to the feature, and I'm not @cadamswaite.

**Comment by cadamswaite** [comment]
If I put the code into a branch of this project, would you be happy to update it to get it through review etc?

**Comment by jayaddison** [comment]
Yep, certainly can - if you could open a pull request with the changes then we can shepherd it from there (I'd add whatever settings/config approach I think is appropriate, for example - no need for you to manage that unless you'd like to) :+1:
